### PR TITLE
daslang -exe: 3-tier shared-module resolution (exe_dir → das_root → absolute)

### PR DESCRIPTION
**Pull request by borisbat**
## Summary

Standalone executables built with `daslang -exe` now resolve dynamic shared modules at startup by trying:

1. `<exe_dir>/<rel_path>` — daspkg release bundles (modules sit next to exe)
2. `<das_root>/<rel_path>` — SDK install AND local dev build
3. `<fallback_abs_path>` — baked-at-codegen absolute (legacy fallback)

Previously only tier 3 existed: the absolute path captured at compile time was baked verbatim into the binary. This worked for in-place dev builds (the path stays valid as long as the exe doesn't move) but **broke two real scenarios** that ship today:

| Scenario | Layout | Before | After |
|---|---|---|---|
| Local dev build | exe at `<repo>/bin/<cfg>/`, modules at `<repo>/modules/` | ✅ baked absolute | ✅ tier 2 (`getDasRoot()` strips `bin/<cfg>` → `<repo>`) |
| CMake install / CI SDK download | exe at `<install>/bin/`, modules at `<install>/modules/` | ❌ baked absolute → build-machine FS | ✅ tier 2 (`<install>/modules/...`) |
| daspkg release bundle | exe at `<bundle>/`, modules at `<bundle>/modules/` | ❌ same | ✅ tier 1 (`<bundle>/modules/...`) |

This is the prerequisite for the upcoming `daspkg release` command (PR #2 of the daspkg release plan) and also fixes the SDK-install case independently.

## Implementation

- `src/builtin/module_jit.cpp` — new `jit_register_dynamic_module_resolve(rel_path, fallback_abs_path, mod_name)` runtime helper. Plus test seams: `jit_set_exe_file_for_test_` and `jit_set_path_exists_for_test_` so the resolution logic is unit-testable without dlopening anything.
- `modules/dasLLVM/daslib/llvm_exe.das` — `inject_main` emits the new helper instead of `jit_register_dynamic_module`. New private `compute_modules_relative_suffix` extracts the `modules/<X>/...` suffix at codegen time via `to_generic_path` + substring search (cross-platform; never scans for both `/X/` and `\X\`).

## Test plan

- [x] `tests-cpp/small/test_jit_module_resolve.cpp` — 7 cases covering resolution priority, _debug-variant detection, Windows backslash exe paths, empty-input edge cases (uses test seams to drive logic with synthetic exe paths and a mock filesystem predicate).
- [x] `tests/exe-paths/_uses_sqlite.das` + `run_layouts.cmake` — end-to-end integration: builds a standalone exe via `daslang -exe`, deploys it to bundle / sdk-install / local-dev layouts in tmp dirs, runs each from a neutral cwd. Labeled `small` so CI's `ctest -L small` step (`build.yml:340/343`) picks it up across the full platform matrix.
- [x] Existing utils (`aot.exe`, `daspkg.exe`, `benchctl.exe`, `mcp.exe` — all built via `daslang -exe`) re-run cleanly after the change. Verified locally.
- [x] `dastest --test tests/`: 7822 tests, 7816 passed, 0 failed, 0 errors, 6 skipped.
- [x] `test_aot --use-aot --test tests/`: 7216 tests, 7210 passed, 0 failed, 0 errors, 6 skipped.

## Doc note

`skills/filesystem.md` — clarifies that the existing "never `rfind('/')` / `rfind('\\\\')` on paths" rule has one legitimate exception: searching for a **named component** (e.g. `/modules/`, `/.git/`) after `to_generic_path` normalize. Single-line addition + one row in the "pick the right tool" table.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: modules/dasLLVM/daslib/llvm_exe.das [hunk 1]
```diff
@@ -559,6 +559,14 @@ def collect_external_functions(standalone_context : LLVMOpaqueValue?; ctx : LLVM
     return (any_unimplemented = any_unimp, needs_whole_lib = whole_lib)
 }
 
+// Suffix from the last "modules/" segment onwards; "" → caller falls back to abs path.
+// to_generic_path normalizes Windows backslashes so the search needs only one form.
+def private compute_modules_relative_suffix(p : string) : string {
+    let pn = to_generic_path(p)
+    let idx = rfind(pn, "/modules/")
+    return idx < 0 ? p : slice(pn, idx + 1)
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasLLVM/daslib/llvm_exe.das]: Real bug — fixed. `compute_modules_relative_suffix` now returns `""` when no `/modules/` segment is found. The C++ resolver already short-circuits on empty `rel_path` (verified by the existing `"empty rel_path skips tier 1+2 and uses absolute"` unit test), so paths without a `/modules/` anchor go straight to tier 3 (baked absolute) — no more `<exe_dir>/<abs>` / `<das_root>/<abs>` accidental matches.

File: src/builtin/module_jit.cpp [hunk 2]
```diff
@@ -1043,6 +1044,102 @@ DAS_API void * jit_register_dynamic_module ( const char * path, const char * mod
...
+#ifndef NDEBUG
+    // CMake debug builds emit <name>_debug.shared_module; if that variant exists,
+    // return the non-debug name and let register_dynamic_module apply its own
+    // rewrite (we mustn't double-rewrite).
+    auto pos = candidate.rfind(".shared_module");
+    if ( pos != das::string::npos && pos + 14 == candidate.size() ) {
+        das::string dbg = candidate.substr(0, pos) + "_debug.shared_module";
+        if ( jit_path_exists(dbg.c_str()) ) return candidate;
+    }
+#endif
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/builtin/module_jit.cpp]: Real bug — fixed. Switched the `_debug.shared_module` existence check from `#ifndef NDEBUG` to `#ifndef DAS_NO_ASSERTIONS` to match `register_dynamic_module`'s rewrite gate (`module_builtin_fio.cpp:1099`). Comment updated to reference the matching site so future edits don't drift again. Catches the `-DDAS_NO_ASSERTIONS=0` Release build divergence you flagged.

File: tests/exe-paths/run_layouts.cmake [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+# Inputs (passed via -D on the cmake -P invocation):
+#   DASLANG       — path to daslang binary
+#   SRC_DAS       — path to uses_sqlite.das (the test program)
+#   SQLITE_MOD    — path to dasModuleSQLITE(.shared_module|*_debug.shared_module)
+#   WORKDIR       — temporary working directory (test-specific)
+#
+# The script compiles SRC_DAS to a standalone exe, then drives it through:
+#   1. Bundle layout    (exe_dir/modules/dasSQLITE/...)
+#   2. SDK install layout (das_root: exe at <root>/bin/, modules at <root>/modules/)
+#   3. Bare layout (no modules dir at all) — expected to fail or silently no-op
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests/exe-paths/run_layouts.cmake]: Comment fixed to match what's actually implemented — three positive scenarios (bundle / sdk_install / local_dev), one per resolution tier. Skipped adding the negative case for now: with tier 3 fallback active, distinguishing "broken bundle" from "ok bundle" requires runtime override machinery (env-var to disable `getDasRoot` lookup or a build flag to skip baked absolute) that doesn't exist yet. Noted that limitation in the comment too.

File: tests/exe-paths/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# Integration test for daslang -exe's 3-tier dynamic-module resolution.
+# See run_layouts.cmake for the per-scenario logic.
+
+if(NOT TARGET daslang OR NOT TARGET dasModuleSQLITE)
+    return()
+endif()
+
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests/exe-paths/CMakeLists.txt:9]: Real bug — fixed. Added `DAS_LLVM_INCLUDED` to the early-return guard. Note: dasLLVM doesn't expose a CMake target (it's a daslang module loaded via `.das_module`, not a `add_library` target), so `if(NOT TARGET dasLLVM)` would always trigger return; `DAS_LLVM_INCLUDED` is the variable [modules/dasLLVM/CMakeLists.txt:2](https://github.com/GaijinEntertainment/daScript/blob/master/modules/dasLLVM/CMakeLists.txt#L2) sets after passing its enabled-check, which is the right signal here.

File: src/builtin/module_jit.cpp [hunk 5]
```diff
@@ -1043,6 +1044,104 @@ DAS_API void * jit_register_dynamic_module ( const char * path, const char * mod
...
+static bool jit_path_exists ( const char * p ) {
+    return g_jit_path_exists_for_test ? g_jit_path_exists_for_test(p) : das::builtin_fexist(p);
+}
+
+// Try a candidate dir + rel_path; if it (or its _debug variant in non-NDEBUG
+// builds) exists, return the path to load. Empty string = miss.
+//
+// Returns the non-_debug name even when the _debug variant is what exists —
+// register_dynamic_module applies the _debug rewrite itself in non-NDEBUG
+// builds, so we mustn't double-rewrite.
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/builtin/module_jit.cpp]: Comment-vs-code drift — fixed. Updated the leading `pick_at_dir` docstring to say "debug builds, gated on DAS_NO_ASSERTIONS to match register_dynamic_module's rewrite" so the comment names the actual macro and points at the loader site we're staying in sync with.

File: modules/dasLLVM/daslib/llvm_exe.das [hunk 6]
```diff
@@ -658,21 +666,27 @@ def public inject_main(program_context : Context?; ctx : LLVMContextRef;
...
+            types.LLVMVoidPtrType(), // const char * fallback_abs_path
             types.LLVMVoidPtrType()  // const char * mod_name
         )
     )
-    var jit_reg_dyn_mod = LLVMAddFunctionWithType(g_mod, "jit_register_dynamic_module", reg_dyn_mod_type)
+    var jit_reg_dyn_mod = LLVMAddFunctionWithType(g_mod, "jit_register_dynamic_module_resolve", reg_dyn_mod_type)
     for_each_registered_dynamic_module() <| $(path, mod_name, das_name) {
         if (key_exists(used_modules, das_name)) {
             has_cpp_modules = true
             to_log(LOG_INFO, "LLVM EXE: include dynamic module: `{das_name}`\n")
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasLLVM/daslib/llvm_exe.das:689]: Real bug — fixed. Switched the second registration site (`collect_external_functions`, the `initialize_modules()` codegen) to also use `jit_register_dynamic_module_resolve`. Both call sites in a standalone exe now route through the same 3-tier resolution, so each module loads exactly once from one chosen path — no more bundle-vs-source-tree double load. JIT-mode behavior is preserved: tier-2 (`<das_root>/modules/<X>/...`) resolves to `<repo>/modules/<X>/...` since `getDasRoot` strips `bin/`.

Caught by your review — well spotted. Force-push: `df5c032f8`.

File: src/builtin/module_jit.cpp [hunk 7]
```diff
@@ -1043,6 +1044,105 @@ DAS_API void * jit_register_dynamic_module ( const char * path, const char * mod
...
+    return true;
+}
+
+// Resolve and load a dynamic module. Used by standalone exes (emitted by
+// inject_main in llvm_exe.das).
+DAS_API void * jit_register_dynamic_module_resolve ( const char * rel_path,
+                                                     const char * fallback_abs_path,
+                                                     const char * mod_name ) {
+    das::string chosen = resolve_dynamic_module_path(rel_path, fallback_abs_path);
+    return das::register_dynamic_module(chosen.c_str(), mod_name, 0/*Quiet*/, nullptr, nullptr);
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/builtin/module_jit.cpp:1144]: Real bug, but pre-existing — the same `nullptr` Context+LineInfoArg pattern is in the existing `jit_register_dynamic_module` ([module_jit.cpp:1042](https://github.com/GaijinEntertainment/daScript/blob/master/src/builtin/module_jit.cpp#L1042)), and only the dlopen-failed branch in `register_dynamic_module` actually respects `Quiet`; the registrator-missing and build-id-mismatch branches do unconditional `context->to_err(...)`.

Filed as #2580 to keep this PR's scope tight on the resolution-tier fix. Will land as a separate small PR that guards the two unconditional branches with the same Quiet check (and/or a defensive `context==nullptr` guard).

File: src/builtin/module_jit.cpp [hunk 8]
```diff
@@ -1043,6 +1044,105 @@ DAS_API void * jit_register_dynamic_module ( const char * path, const char * mod
...
+    das::string exeFile;
+    if ( g_jit_exe_file_for_test ) {
+        const char * s = g_jit_exe_file_for_test();
+        if ( s ) exeFile = s;
+    } else {
+        exeFile = das::getExecutableFileName();
+    }
+    if ( !exeFile.empty() ) {
+        auto sep = exeFile.find_last_of("/\\");
+        das::string exeDir = (sep == das::string::npos) ? das::string(".") : exeFile.substr(0, sep);
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/builtin/module_jit.cpp]: Real bug — fixed by switching to `std::filesystem::path` throughout the resolver. `parent_path()` correctly preserves the filesystem root: POSIX `/myapp` → `/`, Windows `C:\myapp.exe` → `C:\`. Path-joining via `operator/` also handles trailing-separator and root cases (no more `//modules/...` from a root parent). Replaced manual `find_last_of("/\\")` + `substr` and the `dir + "/" + rel_path` concat. Force-push: `59519f8bc`.

File: tests-cpp/small/test_jit_module_resolve.cpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+TEST_CASE("jit module resolve — Windows-style backslash exe path is handled") {
+    TestSeams seams;
+    g_exe_file = "C:\\bundle\\myapp.exe";
+    // Match what pick_at_dir would produce: dirname uses last "/" or "\".
+    g_existing.insert("C:\\bundle/modules/dasFoo/dasFoo.shared_module");
+    auto chosen = resolve("modules/dasFoo/dasFoo.shared_module",
+                          "C:\\build\\abs\\modules\\dasFoo\\dasFoo.shared_module");
+    CHECK(chosen == "C:\\bundle/modules/dasFoo/dasFoo.shared_module");
+}
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests-cpp/small/test_jit_module_resolve.cpp:165]: Tests expanded as part of the same fix. Added:
- POSIX root: `/myapp` exe → `/modules/dasFoo/...` resolves correctly (cross-platform; std::filesystem treats `/` as a separator on every host)
- Windows backslash subdir: `C:\bundle\myapp.exe` → `C:\bundle\modules\dasFoo\...` (gated `#ifdef _WIN32` since std::filesystem::path on POSIX treats `\` as a regular character)
- Windows drive root: `C:\myapp.exe` → `C:\modules\dasFoo\...` (the case you specifically called out — the previous `substr` would have produced `C:` and built malformed `C:modules\...`; now `parent_path()` correctly preserves `C:\`)

The cross-platform "Windows-style backslash" test that was here was a fiction — manual `find_last_of("/\\")` made it appear platform-portable, but that platform-agnostic separator parsing was itself part of the bug. Replaced with the platform-correct setup above.

File: tests-cpp/small/test_jit_module_resolve.cpp [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    g_exe_file = "/somewhere/myapp";
+    // exe-relative does NOT exist; das_root path DOES (use the dev machine's
+    // real das_root so existence check passes against our fake predicate)
+    auto dasRoot = das::getDasRoot();
+    auto dasRootCandidate = dasRoot + "/modules/dasFoo/dasFoo.shared_module";
+    g_existing.insert(dasRootCandidate);
+    auto chosen = resolve("modules/dasFoo/dasFoo.shared_module",
+                          "/build/abs/modules/dasFoo/dasFoo.shared_module");
+    CHECK(chosen == dasRootCandidate);
+}
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests-cpp/small/test_jit_module_resolve.cpp:84]: Real bug — fixed. I introduced this in round 4 by switching to `std::filesystem::path` without realizing `path::string()` emits backslashes on Windows. Both ends now normalize to forward-slash form via `generic_string()`:

- `fake_path_exists`: probes `std::filesystem::path(p).generic_string()` so the resolver's native-form path lookup matches `g_existing` entries written in forward-slash form.
- `resolve()` test helper: returns `std::filesystem::path(buf).generic_string()` so `CHECK(chosen == "/...")` is host-agnostic.

Test inputs and expectations stay readable (forward slashes everywhere). Windows-specific tests updated their expected strings to forward-slash form too (`C:/bundle/...` instead of `C:\\bundle\\...`); the `g_exe_file` raw input still uses backslashes since that's what `getExecutableFileName()` produces on Windows.

The Windows CI matrix entries that were still pending will now exercise the corrected paths. Force-push: `3bf038c7b`.

File: skills/filesystem.md [hunk 11]
```diff
@@ -186,7 +187,7 @@ Fields: `capacity`, `free`, `available` (all `uint64`).
 
 ## Common gotchas
 
-- **Never split paths with `rfind("/")` and `rfind("\\")` followed by `slice`.** Always `base_name` / `dir_name` / `parent`. The manual form misses Windows backslashes, mishandles trailing separators, and returns the wrong slice when no separator is present.
+- **Never split paths with `rfind("/")` and `rfind("\\")` followed by `slice`.** Always `base_name` / `dir_name` / `parent`. The manual form misses Windows backslashes, mishandles trailing separators, and returns the wrong slice when no separator is present. Exception: when the substring you're searching for is a **named component** (e.g. `/modules/`, `/.git/`), normalize once via `to_generic_path(p)` then `rfind(pn, "/modules/")` on the normalized form — the normalize step is what makes the rule "search for `/` only" safe; never search for both `/X/` and `\X\` separately.
```
[the remark this answers is not in the document]
Comment (borisbat) [reply skills/filesystem.md]: Fixed — replaced the phantom `pn` with an inline `to_generic_path(p)` call: `rfind(to_generic_path(p), "/modules/")`. One fewer name, copy-pasteable.
